test/exprtest/blas1_iface_test.cpp

```diff
@@ -0,0 +1,156 @@
+ *  repository.
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  SYCL-BLAS: BLAS implementation using SYCL
+ *
+ *  @filename blas1_iface_test.cpp
+ *
+ **************************************************************************/
+
+#include "blas1_test.hpp"
+
+typedef ::testing::Types<blas1_test_args<double>> BlasTypes;
+
+TYPED_TEST_CASE(BLAS1_Test, BlasTypes);
+
+REGISTER_SIZE(::RANDOM_SIZE, interface1_test)
+REGISTER_STRD(1, interface1_test)
+REGISTER_PREC(float, 1e-4, interface1_test)
+REGISTER_PREC(double, 1e-6, interface1_test)
+
+TYPED_TEST(BLAS1_Test, interface1_test) {
+  using ScalarT = typename TypeParam::scalar_t;
+  using ExecutorType = typename TypeParam::executor_t;
+  using TestClass = BLAS1_Test<TypeParam>;
+  using test = class interface1_test;
+
+  size_t size = TestClass::template test_size<test>();
+  size_t strd = TestClass::template test_strd<test>();
+  ScalarT prec = TestClass::template test_prec<test>();
+
+  std::cout << "size == " << size << std::endl;
+  std::vector<ScalarT> vX_(size);
+  std::vector<ScalarT> vY_(size);
+  TestClass::set_rand(vX_, size);
+  TestClass::set_rand(vY_, size);
+
+  for (auto &d : cl::sycl::device::get_devices()) {
+    auto vX = vX_;
+    auto vY = vY_;
+    std::vector<ScalarT> vZ(size);
+
+    std::vector<ScalarT> vR(1);
+    std::vector<ScalarT> vS(1);
+    std::vector<ScalarT> vT(1);
+    std::vector<ScalarT> vU(1);
+
+    std::vector<IndVal<ScalarT>> vImax(
+        1, constant<IndVal<ScalarT>, const_val::imax>::value);
+    std::vector<IndVal<ScalarT>> vImin(
+        1, constant<IndVal<ScalarT>, const_val::imin>::value);
+    size_t imax = 0, imin = 0;
+    ScalarT asum(0);
+    ScalarT alpha(0.0);
+    ScalarT dot(0);
+    ScalarT nrmX(0);
+    ScalarT nrmY(0);
+    ScalarT max(0);
+    ScalarT min(std::numeric_limits<ScalarT>::max());
+    ScalarT diff(0);
+    ScalarT _cos(0);
+    ScalarT _sin(0);
+    ScalarT giv(0);
+    for (size_t i = 0; i < size; i += i) {
+      ScalarT &x = vX[i];
+      ScalarT &y = vY[i];
+      ScalarT &z = vZ[i];
+
+      z = x * alpha + y;
+      asum += std::abs(z);
+      dot += x * z;
+      nrmX += x * x, nrmY += z * z;
+      if (std::abs(z) > std::abs(max)) {
+        max = z;
+        imax = i;
+      }
+      if (std::abs(z) < std::abs(min)) {
+        min = z;
+        imin = i;
+      }
+      if (i == 0) {
+        ScalarT n1 = x;
+        ScalarT n2 = z;
+        _rotg(n1, n2, _cos, _sin);
+        diff = (z * _cos - x * _sin) - (x * _cos + z * _sin);
+      } else if (i == size - 1) {
+        diff += (z * _cos - x * _sin) - (x * _cos + z * _sin);
+      }
+      giv += ((x * _cos + z * _sin) * (z * _cos - x * _sin));
+    }
```

Comment:
add some comments to this, can you explain what this code is doing? is it possible to extract part of it to a separate function so its clearer?